File: src/cpp_launch_example/src/spiral_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "turtlesim/msg/pose.hpp"

using std::placeholders::_1;

class SpiralPublisher : public rclcpp::Node {
public:
    SpiralPublisher() : Node("spiral_publisher"), distance(0.05) {
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("/turtlesim1/turtle1/cmd_vel", 10);
        subscription_ = this->create_subscription<turtlesim::msg::Pose>(
            "/turtlesim1/turtle1/pose", 10, std::bind(&SpiralPublisher::pose_callback, this, _1));
        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(100), std::bind(&SpiralPublisher::move_turtle, this));
    }

private:
    void pose_callback(const turtlesim::msg::Pose::SharedPtr msg) {
        current_pose = msg;
    }

    void move_turtle() {
        if (!current_pose) {
            return;
        }

        auto vel_msg = geometry_msgs::msg::Twist();

        vel_msg.linear.x = distance;
        vel_msg.angular.z = 2.0;
        distance += 0.025;

        publisher_->publish(vel_msg);
    }

    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subscription_;
    rclcpp::TimerBase::SharedPtr timer_;
    turtlesim::msg::Pose::SharedPtr current_pose;
    double distance;
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SpiralPublisher>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/py_launch_example/py_launch_example/spiral_publisher.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose

class SpiralPublisher(Node):
    def __init__(self):
        super().__init__('spiral_publisher')
        self.publisher_ = self.create_publisher(Twist, '/turtlesim1/turtle1/cmd_vel', 10)
        self.create_subscription(Pose, '/turtlesim1/turtle1/pose', self.pose_callback, 10)
        self.create_timer(0.1, self.move_turtle)
        self.distance = 0.05
        self.current_pose = None

    def pose_callback(self, msg):
        self.current_pose = msg

    def move_turtle(self):
        if self.current_pose is None:
            return

        vel_msg = Twist()

        vel_msg.linear.x = self.distance
        vel_msg.angular.z = 2.0
        self.distance += 0.025

        self.publisher_.publish(vel_msg)

def main(args=None):
    rclpy.init(args=args)
    spiral_publisher = SpiralPublisher()
    rclpy.spin(spiral_publisher)
    spiral_publisher.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
